### PR TITLE
Implement CM Janus synchronization

### PR DESCRIPTION
**Pull request by fvovan**
Implement synchronize method on CM_Janus_Service which uses cm-janus status to add/remove channels and their streams.
This method should be called within maintenance cli when service configured.

### CODE REVIEW

File: library/CM/Janus/Service.php [hunk 1]
```diff
@@ -20,7 +20,64 @@ public function __construct(CM_Janus_Configuration $configuration, CM_Janus_Http
...
+            $streamPublish = $streamChannel->getStreamPublishs()->getItem(0);
+            if ($streamPublish) {
+                if ($streamPublish->getStart() > $startStampLimit) {
+                    continue;
+                }
+                if (!array_key_exists($streamChannel->getKey(), $channelSubscriberList)) {
+                    $streamRepository->removeStream($streamPublish);
+                }
+            } else {
+                //TODO Is it possible to have channel w/o publishers?
```
Comment (fvovan) [inline library/CM/Janus/Service.php]: @tomaszdurka wdyt?

**Comment by fvovan** [comment]
@tomaszdurka  please review,
generally it works but i'm thinking about edge cases
also we need either change output format in `/status` API method at cm-janus
or convert corresponding param names in CM's JanusHttpClient

File: library/CM/Janus/Service.php [hunk 2]
```diff
@@ -20,7 +20,64 @@ public function __construct(CM_Janus_Configuration $configuration, CM_Janus_Http
     }
 
     public function synchronize() {
-        throw new CM_Exception_NotImplemented();
+        $streamRepository = $this->getStreamRepository();
+
+        $startStampLimit = time() - 3;
+        $streamList = [];
+        $channelSubscriberList = [];
+        foreach ($this->_configuration->getServers() as $server) {
```
Comment (tomaszdurka) [inline library/CM/Janus/Service.php]: Oh I see you've copied synchronization from wowza. I think you can rather use `CM_Janus_Service-> _fetchStatus()`

Comment (fvovan) [inline library/CM/Janus/Service.php]: Actually, not entirely. :) Yes, I saw this method, but as you can see there is some processing inside the loop.
I can fetch status using existing method but anyway will have to iterate through it once more.

**Comment by fvovan** [comment]
it's adjusted to use `_fetchStatus`

**Comment by tomaszdurka** [comment]
@fvovan can you double-check what we wrote, plus run tests with coverage.

**Comment by tomaszdurka** [comment]
Change `CM_Janus_HttpApiClient` to expect `{id: stream.id, channelName: stream.channelName, isPublish: stream.isPublish};` response.

**Comment by fvovan** [comment]
@tomaszdurka please review
I fixed some flaws and added several test cases.
Line coverage was increased from 84% to  96%  (basically only getConfigure() left untested, but it is very trivial)

File: library/CM/Janus/Service.php [hunk 3]
```diff
@@ -20,7 +20,55 @@ public function __construct(CM_Janus_Configuration $configuration, CM_Janus_Http
...
+                }
+                if ($streamChannel->getStreamSubscribes()->findKey($clientKey)) {
+                    continue;
+                }
+            }
+            $this->_httpApiClient->stopStream($stream->getServer(), $clientKey);
+        }
+
+        $startStampLimit = time() - 3;
+
```
Comment (tomaszdurka) [inline library/CM/Janus/Service.php]: Remove empty line?

Comment (tomaszdurka) [inline library/CM/Janus/Service.php]: Oh this is on purpose I can see. For each block. Maybe rather not? Probably newline separator for separating stopping client and removing them from database is enough.

**Comment by tomaszdurka** [comment]
Looks good except maybe those newlines. I've restarted failing build.

**Comment by tomaszdurka** [comment]
@fvovan oh, you forgot to add it to maintenance cli.

Please add here https://github.com/cargomedia/cm/issues/1986.
Thanks!

**Pull request merged**